Update: Show all color and gradient origins (core, theme, and user).

jorgefilipecosta opened the pull request:
This PR changes the color components in the block editor inspector and the global styles to show multiple colors and gradient palettes.


The idea of this PR is that the base components ColorPalette and GradientPicker are changed so the color and gradient props are now an array of color sets with the following format: colors = [ { name: "Core", colors: [..] }, { name: "Theme", colors: [..] }, ... ]. That format is optional and only happens when the property __experimentalHasMultipleOrigins is true, so we are totally back-compatible.


## How has this been tested?
I used a background color on the cover block from different origins and verified things worked as expected.
I used a background gradient on the group block from different origins and verified things worked as expected.
I applied a global text, background, and link, color from different origins, and verified things worked as expected.

## Screenshots <!-- if applicable -->

![image](https://user-images.githubusercontent.com/11271197/138929332-ba22a9c4-96bf-43c0-8416-404fda3a8938.png)

![image](https://user-images.githubusercontent.com/11271197/138929634-7f071894-ce31-45de-bcdc-e8395f4090fa.png)


cc: @jasmussen, @pablohoneyhoney would you be able to give a design/UX review to this PR?
cc: @ciampo, @mirka as people components work.

jorgefilipecosta committed:
Update: Show all color and gradient origins (core, theme, and user).

oandregal commented:
Hey, I've given this a run and this is what I see:

### core + theme palette

:ok_hand: I've tried core and theme palette for some themes and it's working well for these ones:

- Empty theme (a theme with theme.json and no color palette)
- TT1-blocks (a theme with theme.json and palette)
- TwentyTwentyOne (theme with theme supports, and color palette, plus link support via theme supports)
- TwentyTwenty (a theme with theme supports and color palette)

:warning: There's this other theme, Storefront, a theme with theme supports and no color palette, for which the theme palette is still shown (although it's the same as core's). 

<img src="https://user-images.githubusercontent.com/583546/139253486-29c58260-88cf-40e3-a24d-0dd36233a94f.png" height="500">

This stems from the fact that `__experimentalFeatures.color.palette.theme` does have the colors as well. This sounds like something to investigate separately. I can look into that.

### user styles

:question: For user palette, this is what I see when the user changes the color value of the "black" color to set it to red-ish (silly thing, I know, but it works for stress testing :) ):

<img src="https://user-images.githubusercontent.com/583546/139253642-3b67084c-aea3-4b03-a8b5-87b203a0d2e8.png" height="500">

:ok_hand:  Then I reset the changes to start over and the user palette is no longer shown.

:question: The other test I've done is what happens when the user adds a new color (reset the changes above and start fresh):

<img src="https://user-images.githubusercontent.com/583546/139253605-7acf8381-8b9c-42c1-b856-9b5997227f38.png" height="500">

:question: Finally, I've switched themes and verified that the user colors are cleared because they are tied to the theme.

This raised a couple of questions around how do we expect this to work for users:

- What do we expect to find in the user color palette? I'd presume only the user modifications (colors whose value was changed or new colors added by the user).
- What do we expect to happen to user color palettes when switching themes?

### blocks that don't use block supports

Without having done a deep dive into the code yet, I've noticed this updates a few components so a number of blocks will have the new design:

- The ones that use the hooks bound to the color block support mechanism.
- Other blocks, such as **cover**, don't use the color support mechanism but they do use the `PanelColorGradientSettings` component from the block-editor package, which was also updated.

I've noticed that there are some other blocks, such as **navigation**, that don't use the color support mechanism and use the `PanelColorSettings` color component from the block-editor package (which wasn't updated, apparently). These won't have the new design. I wonder what's the way forward for these blocks. Sorry if I'm asking questions that have been probably already discussed in the components PRs, but I'm lacking tons of context here.

oandregal commented:
> There's this other theme, Storefront, a theme with theme supports and no color palette, for which the theme palette is still shown

This bug is independent from this PR and I've got a fix at https://github.com/WordPress/gutenberg/pull/36054

jorgefilipecosta commented:
This PR got some conflicts, and I rebased it.

Hi @oandregal,

Thank you for the review!

What this PR does is show the three-color origins when available. The strange behavior noted when changing a theme color or adding a user color is because the color palette editor assumes the default value of user colors are the theme colors, so the duplication happens. We are changing that logic; the user colors will be an empty pallet at the start in https://github.com/WordPress/gutenberg/pull/35783. So after https://github.com/WordPress/gutenberg/pull/35783 is merged, no more duplication is going to happen.
I guess right now, we can leave that issue for https://github.com/WordPress/gutenberg/pull/35783. This PR seems to be doing the right things shows the three available origins as they are stored without any additional logic.

> I've noticed that there are some other blocks, such as navigation, that don't use the color support mechanism and use the PanelColorSettings color component from the block-editor package (which wasn't updated, apparently). These won't have the new design. I wonder what's the way forward for these blocks. Sorry if I'm asking questions that have been probably already discussed in the components PRs, but I'm lacking tons of context here.

Nice catch 👍 PanelColorSettings is updated in the sense that it just uses PanelColorGradientSettings. The issue was that some blocks were not yet passing the flag to enable the multiple origins and is now fixed.

oandregal commented:
> There's this other theme, Storefront, a theme with theme supports and no color palette, for which the theme palette is still shown (although it's the same as core's).

This should have been fixed by https://github.com/WordPress/gutenberg/pull/36054 (this PR would need to be rebased to get those changes). Update: we still need https://github.com/WordPress/gutenberg/pull/36173 as well.

oandregal reviewed (approved):
This works as advertised for me. 

Code-wise, I haven't done a deep dive into this, and I'd appreciate it if people with recent components experience would be able to take a look. Given the time constraints for the freeze, it can also be done as a post-review and refactor later.

We need https://github.com/WordPress/gutenberg/pull/36173 that fixes a recent regression in `trunk` for user colors to work, but that's unrelated to this PR.

oandregal commented:
@jorgefilipecosta I've found a minor thing: when the color selected is white, the color HEX is not visible

![Captura de ecrã de 2021-11-03 11-05-30](https://user-images.githubusercontent.com/583546/140049107-368b0cbb-ba05-4210-b2ab-e75b9af7df81.png)

youknowriad commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -215,3 +216,59 @@ export function getSupportedGlobalStylesPanels( name ) {
 	return supportKeys;
 }
+
+export function useColorsPerOrigin( name ) {
```
I wonder if these would be better on their own "color" file (leaving hooks for the raw hooks only)

youknowriad commented on packages/edit-site/src/components/global-styles/screen-text-color.js:
```diff
@@ -51,8 +58,9 @@ function ScreenTextColor( { name } ) {
-				colors={ solids }
+				colors={ colorsPerOrigin }
 				disableCustomColors={ ! areCustomSolidsEnabled }
+				__experimentalHasMultipleOrigins
```
What would it take to make this the default and remove the prop entirely?

ciampo reviewed (commented):
Hey @jorgefilipecosta , this looks awesome! It's so cool to see the Global Styles UI migrate to the new components!

A bit late to the party, but would it be possible to document the changes made to the `@wordpress/components` package? It'd be good to add at least entries to `packages/components/CHANGELOG.md`.

cc @mirka

ndiego commented:
@jorgefilipecosta is there a way to disable/hide the CORE color palette?

xavisoler commented:
> @jorgefilipecosta is there a way to disable/hide the CORE color palette?

You can follow this issue [#36470](https://github.com/WordPress/gutenberg/issues/36407#issue-1050970557)

ciampo commented:
> Hey @jorgefilipecosta , this looks awesome! It's so cool to see the Global Styles UI migrate to the new components!
> 
> A bit late to the party, but would it be possible to document the changes made to the `@wordpress/components` package? It'd be good to add at least entries to `packages/components/CHANGELOG.md`.
> 
> cc @mirka

Opened #36448 to tackle this follow-up
